### ISSUE TITLE
v1.13.0 - 'ORA-24801: illegal parameter value in OCI lob function' on CLOBs

### ISSUE THREAD

**Issue by MPrtenjak** [oracle/node-oracledb#657]
### For general questions:

Node app using Oracle as backend DB is working without problems on oracledb version 1.12.2, but by changing version to 1.13.0 I am getting random ORA-24801 errors when reading CLOBs.

So everything is the same - Oracle XD DB, same server, same OS, same computer, same app. Just changing oracledb from version 1.12.2 to 1.13.0 I am hitting random ORA-24801.

The problem is not visible on examples (lobstream2.js), but the difference is that my CLOB is returned by function and not directly selected.

#### Answer the following questions:

1. What is your version of Node.js?  

```
Run at: Wed Mar 22 2017 08:40:20 GMT+0100 (Central Europe Standard Time)
Node.js version: v6.10.0
Node-oracledb version: 11300
Node-oracledb text format: 1.13.0
Oracle Client library version: 1102000400
Oracle Client library text format: 11.2.0.4.0
Oracle Database version: 1102000200
Oracle Database text format: 11.2.0.2.0

```

2. What version of node-oracledb are you using?
See answer 1

3. What is the version of your Oracle client (e.g. Instant Client)?  How was it installed?  Where it is installed?
See answer 1

4. What is the version of Oracle Database?
See answer 1

5. What is your OS and version?
Windows Server 2012 R2 Standard

6. What compiler version did you use?  
Visual Studio 2015

7. What environment variables did you set?  
Exactly as described  here: (https://community.oracle.com/docs/DOC-931127)

8. What *exact* command caused the problem (e.g. what command did you try to install with)?  Who were you logged in as?

```
  var bindVars = {
    file      : { dir: oracledb.BIND_OUT, type: oracledb.CLOB },    
    sessionID : sessionID || null,  
    api_key   : api_key || null,
    fileID    : fileID,
  };

  execute("begin :file := GetData(:sessionID, :api_key, :fileID); end;", bindVars, ...
```

This code works perfectly on about 90% of cases, but it randomly throws ORA-24801. Version 1.12.2 works OK in 100% of cases. 

The problem is not related to data in any way. The same CLOB is returned OK, but throws error few seconds later and than works OK in few seconds.

For example I corrected the code so if ORA-24801 is thrown the app wait 500ms and tries again... This corrects most of the problems, but not all of the cases. Sometimes two ORA-24801 are thrown one after another...

I am using 1.12.2 again

9. What error(s) you are seeing?

Randomly "ORA-24801: illegal parameter value in OCI lob function"

**Comment by cjbj** [comment]
Doesn't sound good.  Got a full testcase we can look at?

**Comment by MPrtenjak** [comment]
Hi, I will try to create a minimal test case. But as I said, I didn't get the same results by trying lobstream2.js.

**Comment by MPrtenjak** [comment]
Hi, 

OK; this is the smallest example on which I can reproduce the error. _I know you can write code differently, but this is an example (fully stripped) from production code._

This is the Oracle function:

_The fransiza_interface table is just a normal Oracle table with lots of fields (fransiza_interface_id is the ID and cdata is a clob). In reality the function is much bigger, that is why there are more parameters._

**IMPORTANT: When I was stripping the code to reproduce the problem I discovered that the problem is not visible if I remove the first two parameters (p_sessionId and p_apiKey). I know this is strange, but if I leave them, than the problem is visible, otherwise not?**

```
   FUNCTION MTest (
      p_sessionId               IN NUMBER,
      p_apiKey                  IN VARCHAR,
      p_fransiza_interface_id      NUMBER)
      RETURN CLOB
   IS
      ret_clob    CLOB;
   BEGIN
     SELECT podatki
       INTO ret_clob
       FROM fransiza_interface
      WHERE fransiza_interface_id = p_fransiza_interface_id;

    RETURN ret_clob;
   END; 
```

And this is js file:

```
'use strict';

var oracledb = require('oracledb');
var async = require('async');
var config = require('./config');
	
async.timesSeries(10, function(n, next) {
	console.log(n);
	getFile(next);
}, function(err) {
	if (err)
		console.trace(err);
});
	
function getFile(next) {
  var bindVars = {
    file      : { dir: oracledb.BIND_OUT, type: oracledb.CLOB },    
    sessionID : null,  
    api_key   : null,
		//krneki    : "xx",
    fileID    : 5,
  };

  //_executeSelectInOracle("begin :file := FRAN01.PK_FRANSIZA_API.MTest(); end;", bindVars, 
	_executeSelectInOracle("begin :file := FRAN01.PK_FRANSIZA_API.MTest(:sessionID, :api_key, :fileID); end;", bindVars, 	
    function(err, clobAsString) {
			if (err)
				console.trace(err);
			else
				console.log("result: " + clobAsString.substring(1,20));
			
			if (next) next();
    });
}

function _streamCLOBtoString(connection, clob, callback) {
  var clobAsString = ""; 
  clob.setEncoding('utf8');
  clob.on('error',
    function(err) { return callback(err); });
  clob.on('data',
    function(chunk) { clobAsString += chunk;  });
  clob.on('close',
    function() { doRelease(connection); return callback(null, clobAsString); });
};

function _executeSelectInOracle(selectClause, bindVariables, callback) {
  oracledb.outFormat = oracledb.OBJECT;
  oracledb.getConnection({
    user: config.oracle.user,
    password: config.oracle.password,
    connectString: config.oracle.connectString,
  }, function (err, connection) {
    if (err) {
      console.log("oracle connection error");
      console.log(err);
      callback(err.message);
      return;
    }
    connection.execute(selectClause,
      bindVariables,
      function (err, result) {
        if (err)
          callback(err);
				else
					_streamCLOBtoString(connection, result.outBinds.file, callback);
      });
  });
}

function doRelease(connection) {
  connection.release(
    function (err) {
      if (err) {
        console.error(err);
      }
    }
  );
} 
```

**Comment by MPrtenjak** [comment]
To clarify:

If I change the oracle function like this:

```
   FUNCTION MTest (
      --p_sessionId               IN NUMBER,
      --p_apiKey                  IN VARCHAR,
      p_fransiza_interface_id      NUMBER)
      RETURN CLOB
   IS
      ret_clob    CLOB;
   BEGIN
     SELECT cdata
       INTO ret_clob
       FROM fransiza_interface
      WHERE fransiza_interface_id = p_fransiza_interface_id;

    RETURN ret_clob;
   END;
```

And correct the js file like this:

```
function getFile(next) {
  var bindVars = {
    file      : { dir: oracledb.BIND_OUT, type: oracledb.CLOB },    
    //sessionID : null,  
    //api_key   : null,
    fileID    : 5,
  };

  _executeSelectInOracle("begin :file := FRAN01.PK_FRANSIZA_API.MTest(/*:sessionID, :api_key, */:fileID); end;", bindVars,   
    function(err, clobAsString) {
      if (err)
        console.trace(err);
      else
        console.log("result: " + clobAsString.substring(1,20));
      
      if (next) next();
    });
}
```

The problem is not visible!

**Comment by cjbj** [comment]
Thanks for the testcase.  It does seem to be related to binding NULL to a number.  We'll look at it.

**Comment by cjbj** [comment]
Another comment:  look at using [`fetchAsString`](https://github.com/oracle/node-oracledb/blob/v1.13.0/doc/api.md#propdbfetchasstring) of [`fetchInfo`](https://github.com/oracle/node-oracledb/blob/v1.13.0/doc/api.md#propexecfetchinfo) for CLOBs in place of your `_streamCLOBtoString` code.

**Comment by MPrtenjak** [comment]
Thanks for the comment, but I had a problem with fetchAsString in some cases (something related to max 32000 charh of Varchar). I don't remember the correct version of Oracle DB ~ oracledb, but that is why I am using this dummy but working peace of code :)

**Comment by cjbj** [comment]
Ah yes, you have the 11.2 client.  You wouldn't have that limitation if you used 12.1 or 12.2 - both these connect to your DB version.  Also the future node-oracledb 2.x won't have the limitation with either 11.2 or 12 client libraries.

**Comment by pvenkatraman** [comment]
If the numeric parameter value is null, specify as follows:
sessionID : { type : oracledb.NUMBER, val : null, dir : ORACLEDB.BIND_IN}

**Comment by MPrtenjak** [comment]
As I see it. 

In Oracle any type can be null by design, so if not especially state differently, you can always expect null in any variable.

**Comment by pvenkatraman** [comment]
Please try this as work-around.

**Comment by MPrtenjak** [comment]
O, sorry I thought that you are just making observation, not a suggestion. 

As a "work-around" I have downgraded to 1.12.2 where everything works OK.

**Comment by rafael-pinho** [comment]
Hello

I have the same issue when I tried to execute a stored procedure.  I'm using Oracle Database 11g Release 11.2.0.4.0. I could execute sql statements like select or inserts but, when I tried to execute some stored procedure this error occurs. I saw the problem is when I pass 'null' or 'undefined' as parameter. Then I roolback the node-oracledb package to v1.12.2 and all works fine too.

Sorry the bad english.

Best regards.

**Comment by claudiorodriguez** [comment]
Having the same issue. Everything works just fine with `1.12.2`, but with `1.13.0`, when binding `null` to a parameter to insert into a `VARCHAR`, I get the error.

**Comment by cjbj** [comment]
@pvenkatraman is looking into it.

**Comment by cjbj** [comment]
If I understand the work around correctly, specifying the `type` is what is needed until we release a patch.

**Comment by Ken-Yang** [comment]
We encountered the same issue too. Thanks @MPrtenjak 's workaround. I have downgraded the version to v1.12.2, then everything works fine.

**Comment by cjbj** [comment]
We've just released node-oracledb 1.13.1; it has a patch for this.

The workaround if you have to use 1.13.0 is to explicitly set the `type` when binding.

The problem did not occur in 1.12.2.
